### DISCUSSION TITLE
Local scalar variable for multi-scale simulation

### DISCUSSION THREAD

**Discussion by lingzou** [idaholab/moose#16274] [Q&A General] [answered]
Hi All,

I have a question regarding multi-scale simulation.
Say, I have a field (a mesh), where a field variable lives (say, temperature (T)), whose residual depends on the transient term and gradient term. Normal physics.

In addition to this field variable, I would like to learn some localized physics (say, 'var_local'), whose residual depends on the field variable (its local value, _T[_qp], and gradient, _grad_T[_qp]) and the local variable (some_var), but not its gradient (_grad_some_var[_qp], which does not have physical meaning, neither we should care). Clearly, this 'var_local' lives on each mesh node (and q point).

This 'var_local', depending on user request, could be a single value, or an array. This seems to me should be handled by scalar variable type of approach, but current scalar variable does not seem to support this type of specialized request.

Any suggestions/discussions?

**Comment by aeslaughter** [comment, marked as answer]
Are you creating a Kernel object that contributes to the residual of T that includes contributions that depend on 'var_local'? And 'var_local' can be either a single value or a vector of values, depending on user options? Can you calculate 'var_local' directly given T and grad T? 

If those are all yes, then 'var_local' probably just needs to be a material property. The Kernel can decided which version of the property to use (scalar vs vector) based on an input parameter.

**Reply by lingzou** [reply]
Andrew, thanks for your reply. Sorry that I didn't get back to this reply earlier, I am still more used to watch the email list.

Back to your reply, no, this is not what I would like to implement. 'var_local' is indeed a variable (or variable array), and it has its own residual form. I have explored what currently moose provides and found similar implementations (e.g., ArrayKernal and NodalKernel), and it seems to me that a new class (e.g., ArrayNodalKernal) will do the job. I will work on it, and see how it goes, and maybe in a later time contribute it back to moose framework.
